plugins/net_plugin/net_plugin.cpp

```diff
@@ -3428,10 +3657,45 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
+      //
+      // Check the version before taking the lock.  Since this is the only thread that
+      // touches the version information, no need to take the lock if the version is
+      // unchanged.
+      //
+      auto& update = bs->get_security_group_info();
+      if(security_group.current_version() == update.version) {
+         return;
+      }
+      std::lock_guard<std::shared_mutex> connection_guard(connections_mtx);
+      if(!security_group.update_cache(update.version, update.participants)) {
+         return;
+      }
+
+      // update connections
+      //
+      fc_dlog( logger, "SecurityGroup changed to version: ${v}", ("v", security_group.current_version()) );
+      for(auto& connection : connections) {
+         const auto& participant = connection->participant_name();
+         if(participant && security_group.is_in_security_group(participant.value())) {
+            if(!connection->is_participating()) {
+               connection->set_participating(true);
+               connection->send_handshake();
```

Comment:
Can send_handshake throw? If so, it is better to do set_participating(true) after send_handshake().

Reply:
It doesn't throw, and since it ends up processing that send on another thread, you want participating already set true, otherwise you may end up not completing the chain of events that sending a handshake was meant to cause.